# Distribuição Geométrica

## A probabilidade de serem necessários n insucessos antes do primeiro sucesso é `P(Y = n) = (1 - p)^n * p` para n = 0, 1, 2, 3, ....
### Um sistema de comunicação repete transmissões até receber confirmação de recebimento. A probabilidade de sucesso em cada tentativa é p = 0.3.

a) Criar uma simulação para gerar amostras da distribuição geométrica  
b) Qual a probabilidade de o primeiro sucesso ocorrer na 4ª tentativa?  
c) Qual a probabilidade de precisar de mais de 6 tentativas?  
d) Determine a média e variância da variável aleatória.  

<a href="https://colab.research.google.com/github/jgocana/Trabalhos-TP-547/blob/master/Trabalho%201/distribuição_geométrica.ipynb" target="_parent">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

In [ ]:
!pip install numpy

In [1]:
"""
Distribuição Geométrica

Define-se a variável aleatória como o número de insucessos antes do primeiro sucesso.

Função de probabilidade:
P(Y = n) = (1 - p)^n * p

Onde:
p = 0.3 (probabilidade de sucesso)
n = 0, 1, 2, ...

O script resolve:
(a) Simulação da distribuição
(b) P(primeiro sucesso na 4ª tentativa)
(c) P(mais de 6 tentativas)
(d) Média e variância
"""

import numpy as np

# ==============================
# Parâmetros
# ==============================
p = 0.3
q = 1 - p
N = 100000  # número de amostras

# ==============================
# Simulação (método da inversa)
# ==============================
def gerar_geometrica(p, N):
    q = 1 - p
    amostras = []
    u = np.random.uniform(0, 1, N)

    for ix in u:
        i = 0
        pr = p
        F = pr

        while ix >= F:
            pr = q * pr
            F += pr
            i += 1

        amostras.append(i)

    return np.array(amostras)

# Gerar amostras
amostras = gerar_geometrica(p, N)

# ==============================
# a) Simulação (já feita acima)
# ==============================
print("a) Amostras geradas:")
print(amostras[:20])  # mostrar só as primeiras

# ==============================
# b) Primeiro sucesso na 4ª tentativa
# ==============================
# "4ª tentativa" → 3 falhas antes → Y = 3

p_teorico_b = (q**3) * p
p_simulado_b = np.mean(amostras == 3)

print("\nb) P(primeiro sucesso na 4ª tentativa)")
print("Teórico:", p_teorico_b)
print("Simulado:", p_simulado_b)

# ==============================
# c) Mais de 6 tentativas
# ==============================
# mais de 6 tentativas → mais de 5 falhas → Y > 5

p_teorico_c = q**6
p_simulado_c = np.mean(amostras > 5)

print("\nc) P(mais de 6 tentativas)")
print("Teórico:", p_teorico_c)
print("Simulado:", p_simulado_c)

# ==============================
# d) Média e variância
# ==============================
# Teórico
media_teorica = q / p
variancia_teorica = q / (p**2)

# Simulado
media_simulada = np.mean(amostras)
variancia_simulada = np.var(amostras)

print("\nd) Média e variância")
print("Média teórica:", media_teorica)
print("Média simulada:", media_simulada)
print("Variância teórica:", variancia_teorica)
print("Variância simulada:", variancia_simulada)


a) Amostras geradas:
[6 0 3 0 0 5 3 1 0 0 4 2 0 6 0 0 3 0 1 2]

b) P(primeiro sucesso na 4ª tentativa)
Teórico: 0.10289999999999998
Simulado: 0.10176

c) P(mais de 6 tentativas)
Teórico: 0.11764899999999996
Simulado: 0.11541

d) Média e variância
Média teórica: 2.3333333333333335
Média simulada: 2.31391
Variância teórica: 7.777777777777778
Variância simulada: 7.6373505119


## Resultados  
---  

### a) Amostras geradas

Primeiras amostras obtidas:

[6 0 3 0 0 5 3 1 0 0 4 2 0 6 0 0 3 0 1 2]

---

### b) Probabilidade de sucesso na 4ª tentativa

\[
P(Y = 3)
\]

- Teórico: 0.10289999999999998  
- Simulado: 0.10176  

---

### c) Probabilidade de mais de 6 tentativas

\[
P(Y > 5)
\]

- Teórico: 0.11764899999999996  
- Simulado: 0.11541  

---

### d) Média e variância

- Média teórica: 2.3333333333333335  
- Média simulada: 2.31391  

- Variância teórica: 7.777777777777778  
- Variância simulada: 7.6373505119  
